gpu/gpu_info_oneapi.c

```diff
@@ -0,0 +1,272 @@
+           h.device_info[i].uuid.id[2], h.device_info[i].uuid.id[3],
+           h.device_info[i].uuid.id[4], h.device_info[i].uuid.id[5],
+           h.device_info[i].uuid.id[6], h.device_info[i].uuid.id[7],
+           h.device_info[i].uuid.id[8], h.device_info[i].uuid.id[9],
+           h.device_info[i].uuid.id[10], h.device_info[i].uuid.id[11],
+           h.device_info[i].uuid.id[12], h.device_info[i].uuid.id[13],
+           h.device_info[i].uuid.id[14], h.device_info[i].uuid.id[15]);
+
+  uint32_t memCount = 0;
+  ret = (*h.zesDeviceEnumMemoryModules)(device, &memCount, NULL);
+  if (ret != ZE_RESULT_SUCCESS) {
+    snprintf(buf, buflen, "unable to enumerate Level-Zero memory modules: %d",
+             ret);
+    resp->err = strdup(buf);
+    return;
+  }
+
+  LOG(h.verbose, "discovered %d Level-Zero memory modules\n", memCount);
+
+  zes_mem_handle_t *mems = malloc(memCount * sizeof(zes_mem_handle_t));
+
+  (*h.zesDeviceEnumMemoryModules)(device, &memCount, mems);
+
+  for (m = 0; m < memCount; m++) {
+    zes_mem_state_t state;
+    state.stype = ZES_STRUCTURE_TYPE_MEM_STATE;
+    state.pNext = NULL;
+
+    ret = (*h.zesMemoryGetState)(mems[m], &state);
+    if (ret != ZE_RESULT_SUCCESS) {
+      snprintf(buf, buflen, "unable to get memory state: %d", ret);
+      resp->err = strdup(buf);
+      free(mems);
+      return;
+    }
+
+    resp->total += state.size;
+    resp->free += state.free;
+  }
+
+  free(mems);
+}
+
+ze_result_t oneapi_device_info(oneapi_init_resp_t *resp) {
+  const int buflen = 256;
+  char buf[buflen + 1];
+  int i, d;
+
+  if (resp->oh.handle == NULL) {
+    snprintf(buf, buflen, "Level-Zero handle not initialized");
+    LOG(resp->oh.verbose, "%s\n", buf);
+    return ZE_RESULT_ERROR_UNINITIALIZED;
+  }
+
+  uint32_t driver_count = 0;
+  ze_result_t ret = (*resp->oh.zesDriverGet)(&driver_count, NULL);
+  if (ret != ZE_RESULT_SUCCESS) {
+    snprintf(buf, buflen, "unable to get driver count: %d", ret);
+    LOG(resp->oh.verbose, "%sn", buf);
+    return ret;
+  }
+
+  LOG(resp->oh.verbose, "discovered %d Level-Zero drivers\n", driver_count);
+
+  // get all the driver information
+  zes_driver_handle_t *drivers =
+      (zes_driver_handle_t *)malloc(driver_count * sizeof(zes_driver_handle_t));
+  (*resp->oh.zesDriverGet)(&driver_count, drivers);
+
+  for (d = 0; d < driver_count; d++) {
+    uint32_t device_count = 0;
+    ret = (*resp->oh.zesDeviceGet)(drivers[d], &device_count, NULL);
+    if (ret != ZE_RESULT_SUCCESS) {
+      snprintf(buf, buflen, "unable to get device count: %d", ret);
+      LOG(resp->oh.verbose, "%s\n", buf);
+      free(drivers);
+      continue;
+    }
+
+    LOG(resp->oh.verbose, "discovered %d Level-Zero devices\n", device_count);
+
+    zes_device_handle_t *devices = (zes_device_handle_t *)malloc(
+        device_count * sizeof(zes_device_handle_t));
+    (*resp->oh.zesDeviceGet)(drivers[i], &device_count, devices);
```

Comment:
Just a small hidden bug, the index used here should be `d` instead of `i`. This was causing a segfault on my machine :)

Reply:
That was it, I’ve applied the correction and the segfault I had is now gone.